proposed/2022/dotnet-nuget-why-proposal.md

```diff
@@ -0,0 +1,140 @@
+
+![](../../meta/resources/TransitiveDependencies/TransitiveVSPMUI.png)
+
+Sadly, we don’t have any further CLI or visualizing tool today to answer `Running 'dotnet restore' on a (large) .sln file restores a certain problematic package. How do I know which project(s) in the solution is causing that package to be restored?` question.
+
+## Goals
+The main goal is to add `dotnet nuget why` command to .NET CLI to help customers understand the dependency graph of a given package for PackageReference projects.
+
+Projects that use `packages.config` format are out of scope for this effort because the XML file contains a flat list of both direct and transitive package dependencies. Customers cannot visualize the dependency graph from the file itself but it would be easy to update the package version of transitive depedency incase if there is any security vulnerability in the current version.
+
+## Customers
+We are building this for .NET users to be successful with understanding dependency graph for a given package.
+
+## Solution
+NuGet restore operation generates following files under `obj` folder for `PackageReference` style projects.
+- `project.assets.json` file that maintains a project's dependency graph, which is used to make sure that all necessary packages are installed on the computer.
+- `{projectName}.nuget.dgspec.json` file that maintains a project's top-level dependencies along with other metadata.
+- `{projectName}nuget.g.targets` and `{projectName}.nuget.g.props` files define various properties, such as the path to where the packages cache is on the local machine. They also include any MSBuild imports that NuGet packages referenced by the project need.
+
+The contents of `{projectName}.nuget.dgspec.json` and `project.assets.json` will be helpful to understand the dependency graph of a given package but there are few limitations.
+
+-  NuGet restore operation downloads packages during dependency resolution which may not part of the final graph The absence of a package id and version combination in `project.assets.json` file signals that it was downloaded during dependency resolution but not part of the final dependency graph.
+- Packages acquired through `PackageDownload` are not tied to the project in any way beyond acquisition. These packages are not recorded in the final graph that `project.assets.json` maintains.
+
+`dotnet nuget why` command prints the dependency graph of a given package only if it is part of the final graph.
+
+```
+dotnet nuget why [<PROJECT>|<SOLUTION>] <PACKAGE_NAME>
+    [-f|--framework <FRAMEWORK>]
+    [--version <VERSION>]
+    [-v|--verbosity <LEVEL>]
+
+dotnet nuget why -h|--help
+```
+#### Arguments
+
+- PROJECT | SOLUTION
+
+The project or solution file to operate on. If not specified, the command searches the current directory for one. If more than one solution or project is found, an error is thrown.
```

Comment:
I believe no error should be thrown if more than one project is found.

Reply:
That's a good catch. I guess we can change this statement to if more than one solution or **no** project is found, an error is thrown.

Reply:
> I believe no error should be thrown if more than one project is found.

We should match the behavior of equivalent commands such as `dotnet list package`, `dotnet restore`. 

I think those fail when multiple projects are seen (ambiguity), but best way to figure it out would be just test it out :D

Reply:
> > I believe no error should be thrown if more than one project is found.
> 
> We should match the behavior of equivalent commands such as `dotnet list package`, `dotnet restore`.
> 
> I think those fail when multiple projects are seen (ambiguity), but best way to figure it out would be just test it out :D

Agreed, after testing it out I'll update the spec.

Reply:
After testing out `dotnet list package`, I found that if there were multiple project it would list the packages for each project. So, I will update the spec to reflect similar behavior for the `dotnet nuget why` command.